tac-operation-lifecycle/tac-operation-lifecycle-server/src/services/operations.rs

```diff
@@ -98,37 +114,40 @@ impl TacService for OperationsService {
     ) -> std::result::Result<tonic::Response<OperationsResponse>, tonic::Status> {
         let inner = request.into_inner();
 
-        match inner.q {
-            Some(q) => {
-                let operations = self.db.search_operations(&q).await.map_err(map_db_error)?;
+        let mut input_pagination = None;
+        if let Some(pagination_token) = inner.page_token {
+            input_pagination = Some(LogicPagination {
+                count: inner.page_items.unwrap_or(0) as usize,
+                earlier_timestamp: pagination_token,
+            });
+        }
 
-                Ok(tonic::Response::new(OperationsResponse {
-                    items: OperationsService::convert_short_db_operation_into_response(operations),
-                    next_page_params: None,
-                }))
+        let (operations, pagination) = match inner.q {
+            Some(q) => {
+                // find operations by query
+                self.db
+                    .search_operations(&q, input_pagination)
+                    .await
+                    .map_err(map_db_error)?
             }
-
             None => {
                 // simple operations list with pagination
-                let operations = self
-                    .db
-                    .get_operations(PAGE_SIZE, inner.page_token, OrderDirection::LatestFirst)
+                self.db
+                    .get_operations(input_pagination)
                     .await
-                    .map_err(map_db_error)?;
-
-                let last_timestamp = operations
-                    .last()
-                    .map(|op| op.timestamp.and_utc().timestamp() as u64);
-
-                Ok(tonic::Response::new(OperationsResponse {
-                    items: OperationsService::convert_short_db_operation_into_response(operations),
-                    next_page_params: last_timestamp.map(|ts| Pagination {
-                        page_token: ts,
-                        page_items: inner.page_items.unwrap_or(0) as u32 + PAGE_SIZE as u32,
-                    }),
-                }))
+                    .map_err(map_db_error)?
             }
-        }
+        };
+
+        let next_page_params = pagination.map(|pag| Pagination {
+            page_token: pag.earlier_timestamp,
+            page_items: pag.count as u32,
+        });
```

Comment:
Ideally make `impl From<LogicPagination> for Pagination`, but it will require adding proto as a dependency to logic.

If it's not convenient, maybe make a function in `server` somewhere to do this conversion

Reply:
Yes, since the Pagination is autogenerated struct, i decide to make additional routine `OperationsService::convert_logic_pagination` (https://github.com/blockscout/blockscout-rs/pull/1382/commits/f7500bd87a7970c5a048324ac35a8d51ce585a47)